newsroom/push.py

```diff
@@ -87,17 +121,136 @@ def publish_item(doc):
+        if event.get('state') in [WORKFLOW_STATE.CANCELLED, WORKFLOW_STATE.KILLED] or \
+                event.get('pubstatus') == 'cancelled':
+            # it has been cancelled so don't need to change the dates
+            updates = {'event': event}
+            _id = service.patch(event['guid'], updates)
+        elif event.get('state') in [WORKFLOW_STATE.RESCHEDULED, WORKFLOW_STATE.POSTPONED]:
+            # schedule is changed, recalculate the dates, planning id and coverages from dates will be removed
+            updates = {'event': event}
+            updates['dates'] = get_event_dates(event)
+            _id = service.patch(event['guid'], updates)
+
+    return _id
+
+
+def get_event_dates(event):
+    dates = []
+
+    event['dates']['start'] = datetime.strptime(event['dates']['start'], '%Y-%m-%dT%H:%M:%S+0000')
+    event['dates']['end'] = datetime.strptime(event['dates']['end'], '%Y-%m-%dT%H:%M:%S+0000')
+
+    if event['dates'].get('recurring_rule', None):
+        # recurring event
+        recurring_events = generate_recurring_events(event)
+        for recurring_event in recurring_events:
+            dates.append({
+                'start': recurring_event['dates']['start'],
+                'end': recurring_event['dates']['end'],
+                'state': recurring_event['state'],
+            })
+    else:
+        # non recurring event
+        dates.append({
+            'start': event['dates']['start'],
+            'end': event['dates']['end'],
+            'state': event['state'],
+        })
+    return dates
+
+
+def publish_planning(planning):
+    service = superdesk.get_resource_service('agenda')
+
+    # update dates
+    planning['planning_date'] = datetime.strptime(planning['planning_date'], '%Y-%m-%dT%H:%M:%S+0000')
+
+    if planning.get('event_item'):
+        # this is a planning for an event item
+        agenda = app.data.find_one('agenda', req=None, _id=planning['event_item'])
+
+        if planning.get('state') in [WORKFLOW_STATE.CANCELLED, WORKFLOW_STATE.KILLED] or \
+                planning.get('pubstatus') == 'cancelled':
+
+            # it has been cancelled so remove the planning item from the list
+            agenda['planning_items'] = [p for p in agenda.get('planning_items', []) if p['guid'] != planning['guid']]
+
+            # remove references from dates
+            for date in agenda['dates']:
+                if date['start'].date() == planning['planning_date'].date():
+                    date['planning_id'] = None
+                    date['coverages'] = []
+
+            return service.patch(agenda['_id'], agenda)
+
+        for date in agenda['dates']:
+            if date['start'].date() == planning['planning_date'].date():
+                date['planning_id'] = planning['guid']
+                date['coverages'] = [c['planning']['g2_content_type'] for c in planning.get('coverages', [])]
+
+    else:
+        # there's no event item (ad-hoc planning item)
+        agenda = {}
```

Comment:
here it might search by `planning_id` - there could be updates to ad-hoc planning item

Reply:
Also, you can retrospectively create an event from the ad-hoc planning item.

Reply:
that's a good scenario, not sure what should happen then, I guess there will be another item created for the event and this item should be added to it right @akintolga ?

Reply:
I guess it would update the planning item adding some event_id, so we should get it. but there will be also that event published eventually and I don't think it will have id for the planning item